service-ip-alloc: delay ip processing on service recreate

aroradaman opened the pull request:
Service storage implements transactions. It creates an IPAddress object first and then creates the Service object, and if the Service object already exists the complete transaction is reverted. There can be race conditions when the repair loop picks up the new IPAddress object for reconciliation before the transaction is reverted. This leads to spurious IPAddressWrongReference warnings, to suppress these warnings we delay the processing of the new IPAddress object by 5 seconds. The service allocation creates the IPAddress object before creating the Service object, we easily identify this scenario when Service object is created before the IPAddress object.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind cleanup
/kind bug
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
This PR suppresses the spurious IPAddressWrongReference warning events when existing service is created again.

#### Which issue(s) this PR is related to:
Fixes #132321
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

aroradaman commented:
/cc @aojea @liggitt 
/sig network

liggitt commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -520,6 +520,19 @@ func (r *RepairIPAddress) syncIPAddress(key string) error {
+	// IPAddress object by 5 seconds. The service allocation creates the IPAddress object before creating
+	// the Service object, we easily identify this scenario when Service object is created before the
+	// IPAddress object.
+	if svc.CreationTimestamp.Time.Before(ipAddress.CreationTimestamp.Time) {
```
Don't we only want this to trigger if ipaddress creation timestamp is recent (in the last five seconds) as well?

liggitt replied:
/hold
To prevent accidental merge until ^ is answered

aroradaman replied:
Isn't it safe to assume that transaction in process (which leads to this race) will be reverted. I can only see a requirement of the second check "ipaddress creation timestamp is recent (in the last five seconds)" when transaction which is supposed to be reverted is somehow not reverted.

aroradaman replied:
Ah, in case of kube-apiserver crash the transaction will never be reverted and the repair loop will keep on delaying processing of the IPAddress object indefinitely. Will add that check.

aojea replied:
puff, thanks Jordan,  @aroradaman check the https://github.com/kubernetes/kubernetes/issues/132321#issuecomment-3061763353

aroradaman replied:
I think the second check can be independent of when the service was created. Whenever create is called on existing service the IPAddress will have time.Now() as creation timestamp and we can compare with the clock if it is recent (within 5 seconds..) 

Does the following makes sense?
```go
ipAddress.CreationTimestamp.Time.After(svc.CreationTimestamp.Time) &&
r.clock.Now().Sub(ipAddress.CreationTimestamp.Time) < 5*time.Second
```

aojea replied:
why do you want to make the check undeterministic?

using the object timestamps we have a reliable way to identify the behavior we are worried about

> We just want to handle the case when the transaction revert fails and we don't end up delaying the cleanup of the new IPAddress indefinitely.

we want to avoid the repair loop to race with the transaction, just that

aojea replied:
Oh, right, my bad, is not the service timestamp creation is the timestamp of the duplicate service.... I need to think more about this, we need to avoid we create a loop by constantly requeuing it, that's the goal of this second condition

aroradaman committed:
service-ip-alloc: make repair loop consumer clock interface

The private constructor for the repair loop now consumes clock.Clock interface allowing
predictable unit testing.

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman committed:
service-ip-alloc: delay ip processing on service recreate

Service storage implements transactions. It creates an IPAddress object first and then creates
the Service object, and if the Service object already exists the complete transaction is
reverted. There can be race conditions when the repair loop picks up the new IPAddress object
for reconciliation before the transaction is reverted. This leads to spurious
IPAddressWrongReference warnings, to suppress these warnings we delay the processing of the new
IPAddress object by 5 seconds. The service allocation creates the IPAddress object before creating
the Service object, we easily identify this scenario when the IPAddress object creation timestamp
is after the Service creation timestamp. We do this only when the IPAddress object is created
recently in order to avoid indefinitely requeue/delay in IPAddress cleanup if for some reason
the service transaction revert fails.

Signed-off-by: Daman Arora <aroradaman@gmail.com>

adilGhaffarDev commented:
@liggitt @aojea can we backport it to 1.33?

aojea commented:
This is a cosmetical change , there is no impact more than noisy events on the special case you are recreating a service that already exist ... does not looks it fits into the category of cherry picks https://github.com/kubernetes/community/blob/master/contributors/devel/sig-release/cherry-picks.md#what-kind-of-prs-are-good-for-cherry-picks

xhejtman commented:
Hello,

In our case, the [Dragonfly operator](https://github.com/dragonflydb/dragonfly-operator) particularly triggers this bug, creating nearly 200 000 events for a few instances and doing so faster than we can delete them. Even after attempting to purge events older than one minute, roughly 25 000 events remain:

```
kubectl get events -n default | wc -l
23586
```

This imposes a significant load on both the cluster and the Kubernetes UI. I believe cherry pick should be reconsidered for inclusion in the 1.33 release line.

xhejtman commented:
> > creating nearly 200 000 events for a few instances and doing so faster than we can delete them
> 
> I think that impact on performance is a regression that didn't exist before so it qualifies for cherry pick
> 
> cc: @adilGhaffarDev
> 
> @aroradaman @xhejtman do you mind to submit the cherry pick to 1.33 branch?

I’ll try to tackle this as soon as I have time—definitely this week—unless @aroradaman has already taken care of it.
